### PR TITLE
STM32: FDCAN fix for FIFO1 usage and handling of Error interrupts.

### PR DESCRIPTION
**Pull request by H-Grobben**
FDCAN peripheral was not performing as expected, so I debugged and tested it.

Original code where this was based on used a independent state with regards to the interrupt. During heavy bus error conditions the internal state could become out-of-sync with the interrupts.

Testing on NucleoG474 and using PR #6911 (which is not yet merged..) but checked the HAL function to work with H7 family

More explanation:
During the development of an application using the CAN communication, we found a interrupt-run-away in some situations. We found that the error interrupt triggered (al be it Warning, Passive or Bus-Off, all triggered it) the run-away. We could not get out of the non-responsive state without a reset.

I checked the code, and found two things: the error interrupt is enabled but not cleared in the interrupt routine, and an internal variable 'State' that was used to track the message received state (empty, new, full, overflow) that was not directly related to interrupt that indicated the state.

In this pull request I fixed these issue by adding more values for the interrupt reason (warning, passive, bus off) and clearing the error interrupts, and making the internal state directly dependent on the interrupt state for received messages.

While testing the fixes I also found that introducing the FIFO1 in the CAN receive stage, another issue existed. Even if the messages are received into the FIFO1 (by selecting message filtering for FIFO0 and FIFO1), the interrupt firing was indicating FIFO0 Rx. The configuration of the interrupts for this is now also fixed. The CAN peripheral has 2 interrupt lines going into the NVIC controller. The assignment of the interrupt reasons to these 2 interrupt lines was missing. Now the reception of FIFO1 messages, triggers the 2nd interrupt line. Other interrupts (Rx FIFO0 and Bus Error) are assigned to the 1st interrupt line.

### CODE REVIEW

File: ports/stm32/pyb_can.c [hunk 1]
```diff
@@ -376,8 +387,35 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_1(pyb_can_state_obj, pyb_can_state);
 // Get info about error states and TX/RX buffers
 STATIC mp_obj_t pyb_can_info(size_t n_args, const mp_obj_t *args) {
     #if MICROPY_HW_ENABLE_FDCAN
-    // TODO implement for FDCAN
-    return mp_const_none;
+    // Done: check implementation for FDCAN
```
Comment (dpgeorge) [inline ports/stm32/pyb_can.c:390]: I guess this comment can be removed now, because it works?

File: ports/stm32/make-stmconst.py [hunk 2]
```diff
@@ -65,7 +65,7 @@ class Lexer:
         (
             "#define typedef",
             re.compile(
-                r"#define +(?P<id>[A-Z0-9_]+(ext)?) +\(\([A-Za-z0-9_]+_TypeDef \*\) (?P<id2>[A-Za-z0-9_]+)\)($| +/\*)"
+                r"#define +(?P<id>[A-Z0-9_]+(ext)?) +\(\([A-Za-z0-9_]+_(Global)?TypeDef \*\) (?P<id2>[A-Za-z0-9_]+)\)($| +/\*)"
```
Comment (dpgeorge) [inline ports/stm32/make-stmconst.py:68]: What typedef struct needs this change?

**Comment by iabdalkader** [comment]
Hi, does this PR fix this issue #8169 ?

File: ports/stm32/pyb_can.c (hunk 1 above)
Comment (H-Grobben) [reply ports/stm32/pyb_can.c:390]: Indeed, it can. I will put it on my todo list. Next to the G4 PR...

**Comment by iabdalkader** [comment]
> the problem you are facing is located in the pyb_can.c file, in function pyb_can_send

Yes but FDCAN is initialized in `fdcan.c` in classical CAN mode, it will not support transmission/reception of FD frames. This FD/Classic mode needs to be configurable in `fdcan.c:can_init` and detected in `pyb_can_send()` and `fdcan.c:can_receive()`. Moreover, if understand correctly from the datasheet, it seems possible to configure CAN in FD mode, and control the frame format via FDF bit in TX buffer.

**Comment by iabdalkader** [comment]
FYI I sent this  #8185, it enables FD frames + BRS support, and fixes the Message RAM usage. I also documented how the message RAM is partitioned and what each section is used for. You should probably rebase on that after it's merged. There's minimal changes to `fdcan.c`

**Comment by JohnieBraaf** [comment]
I have tested this PR, it works very well. Also nice to see the `info()` message has been impemented.

**Comment by dpgeorge** [comment]
Thanks for the effort on this @H-Grobben and thanks @JohnieBraaf for testing.

Squashed and merged in 160e4d9a6d491cf3af0843feba5f987b2dd83c60 and 517e82eb6b7d93134637b5658132479f4fc260e3 with some minor clean up (removed duplicate definition of some FIFO macros, and removed duplicated code in `info()`).

**Pull request closed without merging**